modules/economy-system/economy-system.js

```diff
@@ -124,86 +124,317 @@ async function editBank(client, id, action, value) {
+                '%price%': price,
+                '%role%': role.name
+            }, {ephemeral: !interaction.client.configurations['economy-system']['config']['publicCommandReplies']}));
+
+            interaction.client.logger.info(`[economy-system] ` + localize('economy-system', 'created-item', {
+                u: interaction.user.tag,
+                n: name,
+                i: id
+            }));
+            if (interaction.client.logChannel) interaction.client.logChannel.send(`[economy-system] ` + localize('economy-system', 'created-item', {
+                u: interaction.user.tag,
+                n: name,
+                i: id
+            }));
+            await shopMsg(interaction.client);
             resolve(localize('economy-system', 'created-item'));
         }
     });
 }
 
+/**
+ * Function to buy an item
+ * @param {*} interaction Interaction
+ * @param {*} id Id of the item
+ * @param {*} name Name of the item
+ */
+async function buyShopItem(interaction, id, name) {
+    if (!interaction) return;
+    const item = await interaction.client.models['economy-system']['Shop'].findAll({
+        where: {
+            [Op.or]: [
+                {name: name},
+                {id: id}
+            ]
+        }
+    });
+    if (item.length < 1) return interaction.reply({
+        content: interaction.client.configurations['economy-system']['strings']['notFound'],
+        ephemeral: !interaction.client.configurations['economy-system']['config']['publicCommandReplies']
+    });
+    else if (item.length > 1) return interaction.reply({
+        content: interaction.client.configurations['economy-system']['strings']['multipleMatches'],
+        ephemeral: !interaction.client.configurations['economy-system']['config']['publicCommandReplies']
+    });
+
+    if (interaction.member.roles.cache.has(item[0]['role'])) return interaction.reply({
+        content: interaction.client.configurations['economy-system']['strings']['rebuyItem'],
+        ephemeral: !interaction.client.configurations['economy-system']['config']['publicCommandReplies']
+    });
+    let user = await interaction.client.models['economy-system']['Balance'].findOne({
+        where: {
+            id: interaction.user.id
+        }
+    });
+    if (!user) {
+        createUser(interaction.client, interaction.user.id);
+        user = await interaction.client.models['economy-system']['Balance'].findOne({
+            where: {
+                id: interaction.user.id
+            }
+        });
+    }
+    if (user.balance < item[0]['price']) return interaction.reply({
+        content: interaction.client.configurations['economy-system']['strings']['notEnoughMoney'],
+        ephemeral: !interaction.client.configurations['economy-system']['config']['publicCommandReplies']
+    });
+    await editBalance(interaction.client, interaction.user.id, 'remove', item[0]['price']);
+    console.log(item[0]['role']);
+    await interaction.member.roles.add(item[0]['role']);
```

Comment:
I did this so that the user definitely has enough money to buy the item. And I check the perms when creating the item. Therefor it could only lead to errors if the server owner or some admin changes the perms afterwards

Reply:
I was aware that permission checks are being made before the item gets created, but admins often change role positions, permissions and more, sometimes even deleting the role. Adequate handling of such cases is required, as users get charged for a role that they might not receive. The opposite case, where a user loses money after the initial balance check has already been completed, but the role not yet assigned (a race-condition) is more unlikely, but if you want to handle such cases, please add a system that refunds the charge if role assignment fails (adding to balance when the role has not been assigned)